### PR TITLE
cmd/k8s-operator,cmd/containerboot,ipn,k8s-operator: don't run operator's egress proxies with stateful filter

### PR DESCRIPTION
**Pull request by irbekrm**
Turn off stateful filtering for egress proxies to allow cluster
traffic to be forwarded to tailnet.

Allow configuring stateful filter via tailscaled config file.

Deprecate EXPERIMENTAL_TS_CONFIGFILE_PATH env var and introduce a new
TS_EXPERIMENTAL_VERSIONED_CONFIG env var that can be used to provide
containerboot a directory that should contain one or more
tailscaled config files named cap-<tailscaled-cap-version>.hujson.
Containerboot will pick the one with the newest capability version
that is not newer than its current capability version.

Proxies with this change will not work with older Tailscale
Kubernetes operator versions - users must ensure that
the deployed operator is at the same version or newer (up to
4 version skew) than the proxies.

### CODE REVIEW

File: cmd/k8s-operator/sts.go [hunk 1]
```diff
@@ -355,12 +367,12 @@
 	}
 
 	if orig != nil {
-		logger.Debugf("patching the existing proxy Secret with tailscaled config %s", sanitizeConfigBytes(secret.Data[tailscaledConfigKey]))
+		logger.Debugf("patching the existing proxy Secret with tailscaled config %s", sanitizeConfigBytes(secret.Data[tsConfigFileName]))
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/sts.go:370]: False alert, the auth key gets overwritten to '***' for logging purposes, also the logic didn't change in this PR

File: cmd/k8s-operator/sts.go [hunk 2]
```diff
@@ -355,12 +367,12 @@
...
 		if err := a.Patch(ctx, secret, client.MergeFrom(orig)); err != nil {
 			return "", "", err
 		}
 	} else {
-		logger.Debugf("creating a new Secret for the proxy with tailscaled config %s", sanitizeConfigBytes([]byte(secret.StringData[tailscaledConfigKey])))
+		logger.Debugf("creating a new Secret for the proxy with tailscaled config %s", sanitizeConfigBytes([]byte(secret.StringData[tsConfigFileName])))
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/sts.go:375]: False alert, the auth key gets overwritten to '***' for logging purposes, also the logic didn't change in this PR

File: cmd/k8s-operator/sts.go [hunk 3]
```diff
@@ -740,12 +742,13 @@ func enableMetrics(ss *appsv1.StatefulSet, pc *tsapi.ProxyClass) {
...
 // generated and a Secret with the previous proxy state and auth key and
 // returns tailscaled configuration and a hash of that configuration.
-// As of now (9/5/2024 it also returns legacy tailscaled config without the
-// later added NoStatefulFilter field to support proxies older than 1.66.
+//
+// As of 2024-05-09 it also returns legacy tailscaled config without the
+// later added NoStatefulFilter field to support proxies older than cap95.
 // TODO (irbekrm): remove the legacy config once we no longer need to support
-// versions older than 1.66,
+// versions older than cap94,
```
Comment (irbekrm) [inline cmd/k8s-operator/sts.go:749]: 95

File: cmd/k8s-operator/sts.go [hunk 4]
```diff
@@ -524,13 +528,6 @@ func (a *tailscaleSTSReconciler) reconcileSTS(ctx context.Context, logger *zap.S
 			},
 		},
 	}
-	for key := range configs {
-		fn := kubeutils.TailscaledConfigFileNameForCap(key)
-		configVolume.VolumeSource.Secret.Items = append(configVolume.VolumeSource.Secret.Items, corev1.KeyToPath{
-			Key:  fn,
-			Path: fn,
-		})
-	}
```
Comment (irbekrm) [inline cmd/k8s-operator/sts.go]: Removed these after spending a fair bit of time trying to figure out how to test/fix tests. There is really no need to explicitly name them as we use the same names as Secret keys (and we have tests for Secret key names/contents)

File: cmd/k8s-operator/sts.go [hunk 5]
```diff
@@ -770,10 +820,13 @@ type ptrObject[T any] interface {
 // thing that changed is operator version (the hash is also exposed to users via
 // an annotation and might be confusing if it changes without the config having
 // changed).
-func hashBytes(b []byte) (string, error) {
-	h := sha256.New()
-	_, err := h.Write(b)
+func tailscaledConfigHash(c tailscaleConfigs) (string, error) {
+	b, err := json.Marshal(c)
```
Comment (maisem) [inline cmd/k8s-operator/sts.go:824]: You can just do deephash.Hash(c).String()

File: cmd/k8s-operator/sts.go [hunk 6]
```diff
@@ -571,10 +583,6 @@ func (a *tailscaleSTSReconciler) reconcileSTS(ctx context.Context, logger *zap.S
 			VolumeSource: corev1.VolumeSource{
 				Secret: &corev1.SecretVolumeSource{
 					SecretName: proxySecret,
-					Items: []corev1.KeyToPath{{
```
Comment (maisem) [inline cmd/k8s-operator/sts.go:574]: Is this intentional?

**Review by maisem — APPROVED** [review]

File: cmd/k8s-operator/sts.go (hunk 6 above)
Comment (irbekrm) [reply cmd/k8s-operator/sts.go:574]: Re-instated 👍🏼

File: cmd/k8s-operator/sts.go (hunk 5 above)
Comment (irbekrm) [reply cmd/k8s-operator/sts.go:825]: Discussed on Slack, comment above the function as to why we don't use our internal hash here

**Pull request merged**